Limit party size to 30 members

Yutsuten opened the pull request:
_note from Alys to Alys: when notifying the wiki editors about this new feature, include this: https://github.com/HabitRPG/habitica/pull/8589#issuecomment-289715055_

----

[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #8561 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
To-do list:
- [x] Add a field that will contain the information about the current number of members in the party and the limit
- [x] Create a global variable with the maximum value of members in a party
- [x] Hide the button when the limit is reached
- [x] Enforce the limit in the server (API) considering:
  - [x] Current number of members
  - [x] Current pending invitations in the party
  - [x] Number of emails and UUIDs provided for invitations
- [x] Create tests to prove that the limit:
  - [x] Applies to parties
  - [x] Does not apply to guilds

### Current results
Party with 29 members, and trying to invite more 2:

![Invite2](http://i.imgur.com/T4ssO51.png?1)

Result (notification with error):

![Error](http://i.imgur.com/2u9vEwW.png?1)

After inviting only one (button to invite is hidden):

![Invited1](http://i.imgur.com/aNXmv4Z.png?1)

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID: 203f208a-e0f0-4442-bf16-d4f19e553e8a

Yutsuten committed:
Added a field in Party page with members count and maximum members in party

Yutsuten committed:
Added information of invitations counter

Yutsuten committed:
Limited party to 2 members on server (API)

Yutsuten committed:
Fixed english text

Yutsuten committed:
Consider current number of invitations in the party

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -1066,6 +1067,31 @@ api.inviteToGroup = {
 
+    // If party, check the limit of members
+    if (group.type === 'party') {
+      let partyLimitMembers = 5;
```
I'd move this to a `const PARTY_LIMIT_MEMBERS` outside of the function

paglias replied:
actually since the limit is used in the jade file below it should probably be put in common/script/constants so that it can be used in both the server and the client

Yutsuten replied:
Thank you for the feedback! I'll change that.

paglias commented:
This will need tests but it looks okay overall.

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -1066,6 +1067,31 @@ api.inviteToGroup = {
+      if (emails)
+        memberCount += emails.length;
+
+      if (memberCount > partyLimitMembers)
```
let's add parenthesis around the `if`, it matches our code style better

Yutsuten replied:
Do you mean something like this?
``` JavaScript
if (emails) {
   memberCount += emails.length;
}
```

paglias replied:
yes

Yutsuten commented:
@paglias I have no experience with tests, but I would like to try it. Do you know how I could learn and create the necessary tests to validate the changes I'm doing here?

paglias commented:
@mateus-etto all the tests for controllers can be found in `test/api/v3/integration`. You can get a good understanding from there but if you have any question feel free to post here and I'll answer.

The software we use for testing is mocha.js, chai (expect version) and sinon.js

You'll want to test in particular that the limit is enforced with correct error codes, ...

Yutsuten committed:
Moved PARTY_LIMIT_MEMBERS to common folder

Yutsuten committed:
Access the PARTY_LIMIT_MEMBERS through groupsCtrl

Yutsuten committed:
Some corrections

Yutsuten committed:
Hide invite button when invite limit is reached

Yutsuten committed:
Added missing trailing comma

Yutsuten commented:
@paglias I finished my changes in the code, so now I'm going to the tests.

I installed mocha.js, chai and sinon.js, and was able to run ``npm test`` successfully. But the complete test takes about 10 minutes to run, and I would like to run only a single script or test.
I tried ``mocha <path/to/test.js>``, ``mocha --grep <pattern>`` and ``mocha *test*`` but I'm getting a lot of errors...

Error when using path to file: 
"error: Error: Slack url is required"

Error when using ``--grep <patterns>`` (on folder test/api/v3/integration/groups): 
"Warning: Could not find any test files matching pattern: user id invites"

Error when using ``mocha *test*`` (on folder test/api/v3/integration/groups): 
"(function (exports, require, module, __filename, __dirname) { **import** {
SyntaxError: Unexpected token import"

Could you help me?

paglias commented:
@mateus-etto that's great! For tests you can run integration tests only with `gulp test:api-v3:integration` and unit tests with `gulp test:api-v3:unit`. If you want to start the server separately from the integration tests you can do `NODE_ENV=test npm start` and then in a different terminal window `gulp test:api-v3:integration:separate-server`.

To run only a single test or only a single file you'll have to change `describe('section name', ...)` or `it('test name', ...)` to `describe.only(...)` or `test.only(..)`.

Yutsuten committed:
Do not test 'returns only first 30 invites' in a party anymore, but in a guild: party is limited to 30 members, so it would always fail

Yutsuten committed:
Test: allow 30 members in a party

Yutsuten committed:
Test: do not allow 30+ members in a party

Yutsuten committed:
Improved 'allow 30 members in a party' test

Yutsuten committed:
Test: 'allow 30+ members in a guild'

Yutsuten committed:
Added missing trailing comma

Yutsuten commented:
I "fixed" the test "returns only first 30 invites" that was being done in a party. As a party can have up to only 30 members now, I changed the check to a guild instead.

I also added 3 tests:
1. Party with 1 leader and 29 invites, total 30 members (must finish with success)
2. Party with 1 leader and 30 invites, total 31 members (must finish with fail)
3. Guild with 1 leader and 30 invites, total 31 members (must finish with success)

With this, I finished my to-do list for this PR. I'm ready for code/test reviews.

paglias commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -321,6 +322,20 @@ describe('Post /groups/:groupId/invite', () => {
+      expect(await inviter.post(`/groups/${group._id}/invite`, {
+        uuids: generatedInvites.map(invite => invite._id),
+      }))
+      .to.be.an('array');
```
this should be on the previous line

paglias commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -410,5 +425,37 @@ describe('Post /groups/:groupId/invite', () => {
+      expect(await inviter.post(`/groups/${party._id}/invite`, {
+        uuids: generatedInvites.map(invite => invite._id),
+      }))
+      .to.be.an('array');
```
previous line

paglias commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -410,5 +425,37 @@ describe('Post /groups/:groupId/invite', () => {
+      .to.be.an('array');
+    });
+
+    it('do not allow 30+ members in a party', async () => {
```
does not allow

paglias commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -410,5 +425,37 @@ describe('Post /groups/:groupId/invite', () => {
       expect((await userToInvite.get('/user')).invitations.party.id).to.equal(party._id);
     });
+
+    it('allow 30 members in a party', async () => {
```
allows

paglias commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -321,6 +322,20 @@ describe('Post /groups/:groupId/invite', () => {
       });
     });
 
+    it('allow 30+ members in a guild', async () => {
```
allows

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -1066,6 +1067,31 @@ api.inviteToGroup = {
+      let memberCount = 0;
+
+      // Counting the members that already joined the party
+      memberCount += group.memberCount;
```
can you add a blank line after this one? It helps with readability

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -1066,6 +1067,31 @@ api.inviteToGroup = {
+        .find(query)
+        .select('_id')
+        .exec();
+      memberCount += groupInvites.length;
```
blank line

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -1066,6 +1067,31 @@ api.inviteToGroup = {
+      }
+      if (emails) {
+        memberCount += emails.length;
+      }
```
blank line

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -1066,6 +1067,31 @@ api.inviteToGroup = {
+      if (emails) {
+        memberCount += emails.length;
+      }
+      if (memberCount > common.constants.PARTY_LIMIT_MEMBERS)
```
parenthesis

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -1066,6 +1067,31 @@ api.inviteToGroup = {
+        .exec();
+      memberCount += groupInvites.length;
+      // Counting the members that are going to be invited by email and uuids
+      if (uuids) {
```
can you move this if and the one after (`if (email)`) to after the call to `Group.validateInvitations`? Because at this point the uuids and emails params have not been validated and may not be arrays

paglias reviewed (changes requested):
I've let a lot of comments but they're mostly style issues. The code is very good

Yutsuten committed:
Code style corrections

Yutsuten committed:
Fixed new line position

paglias commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -1068,6 +1069,36 @@ api.inviteToGroup = {
+      memberCount += groupInvites.length;
+      
+      // Counting the members that are going to be invited by email and uuids
+      if (uuids) {
```
last edit then I think it's good to go: since we already have a function to validate invites ( `Group.validateInvitations` ) I think we should use that for validating the number of invites too. It's only used here so feel free to change it if you need to pass any other parameter that you might need

Yutsuten replied:
Do you mean that I could put all my code that validates the size of party inside ``Group.validateInvitations``?

paglias replied:
yes, exactly

paglias commented:
Other than my last commit, can you post a screenshot of the changes when you've reached the maximum amount of members?

paglias commented:
Yes

Il 24 Mar 2017 19:28, "Mateus Etto" <notifications@github.com> ha scritto:

> *@mateus-etto* commented on this pull request.
> ------------------------------
>
> In website/server/controllers/api-v3/groups.js
> <https://github.com/HabitRPG/habitica/pull/8589#discussion_r107971017>:
>
> > +      let memberCount = 0;
> +
> +      // Counting the members that already joined the party
> +      memberCount += group.memberCount;
> +
> +      // Count how many invitations currently exist in the party
> +      let query = {};
> +      query['invitations.party.id'] = group._id;
> +      let groupInvites = await User
> +        .find(query)
> +        .select('_id')
> +        .exec();
> +      memberCount += groupInvites.length;
> +
> +      // Counting the members that are going to be invited by email and uuids
> +      if (uuids) {
>
> Do you mean that I could put all my code that validates the size of party
> inside Group.validateInvitations?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/HabitRPG/habitica/pull/8589#discussion_r107971017>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkAVxUmujhTFmm_ODlvouUfhDFGfE8pks5rpAsxgaJpZM4MjRbI>
> .
>

Yutsuten commented:
Something like this?

![InvitedAll](http://i.imgur.com/aNXmv4Z.png?1)

paglias commented:
@mateus-etto that looks good, thanks!

Yutsuten committed:
Party limit check done inside Group.validateInvitations function

mateus-etto commented:
Looks like that changing the ``Group.validateInvitations`` function broke some other tests...
I'll see if I can make it work again.

paglias commented:
They're probably failing because the validation function is now
asynchronous so the tests will need to be changed accordingly

Il 24 Mar 2017 20:42, "Mateus Etto" <notifications@github.com> ha scritto:

> Looks like that changing the Group.validateInvitations function broke
> some other tests...
> I'll see if I can make it work again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/HabitRPG/habitica/pull/8589#issuecomment-289124952>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkAVwZicIahWnlWGvxOAKiZekX-w9NZks5rpByjgaJpZM4MjRbI>
> .
>

Yutsuten committed:
Improved members count query

Yutsuten committed:
Fixed tests

Yutsuten commented:
@paglias I did as you suggested, thank you for the suggestions.
I think I fixed the tests now, locally they worked without problems...

deltaidea commented:
I've been randomly following this conversation over the last couple days and I've gotta say, this is the most friendly and open to contributions project I've ever seen.

paglias commented on test/api/v3/unit/models/group.test.js:
```diff
@@ -460,73 +460,67 @@ describe('Group Model', () => {
       });
 
-      it('throws an error if no uuids or emails are passed in', (done) => {
+      it('throws an error if no uuids or emails are passed in', async () => {
```
there's a problem with these tests, when you have an `async` function and you call something async like `Group.validateInvitations` you need to put `await` in front of it otherwise `Group.validateInvitations` will be called but not waited for and the function will continue to the next step regardless of what happens in `Group.validateInvitations`. So the tests are passing but actually the expec

Also this test was written to make sure that if `Group.validateInvitations ` didn't throw an error (the expected behavior) the test would timeout because `done()` wouldn't be called. Doing the same with `async/await` is less straightforward but you can try something like this

```
      it('throws an error if no uuids or emails are passed in', () => {
       let isSuccessful = false;
        try {
           await Group.validateInvitations(null, null, res);
           isSuccessful = true;
        } catch (err) {
            expect(err).to.be.an.instanceof(BadRequest);
            expect(res.t).to.be.calledOnce;
            expect(res.t).to.be.calledWith('canOnlyInviteEmailUuid');
        });

        if (isSuccessfull) throw new Error('Group.validateInvitations should have throw an error but it did not.')
   })
```

Yutsuten commented:
I see... I tried something different:

``` JavaScript
it.only('throws an error if no uuids or emails are passed in', async () => {
  await expect(Group.validateInvitations(null, null, res)).to.eventually.be.rejected.and.eql({
    httpCode: 400,
    message: 'Bad request.',
    name: 'BadRequest',
  });
  expect(res.t).to.be.calledOnce;
  expect(res.t).to.be.calledWith('canOnlyInviteEmailUuid');
});
```

The test finishes with success, but is this code okay in your opinion?

Yutsuten commented:
I also can change:

``` JavaScript
.to.eventually.be.rejected.and.eql({
    httpCode: 400,
    message: 'Bad request.',
    name: 'BadRequest',
  });
```

to: ``.to.eventually.be.rejected.and.to.be.an.instanceof(BadRequest);``

The code would get even more cleaner. Tested here and looks ok.

paglias commented:
Oh yes, I forgot about that format but it's much more readable. Let's keep it like your first example that is the same we use in all other tests, without the `and.to.be.an.instanceof(BadRequest); `

Yutsuten committed:
Rewrite tests

Yutsuten committed:
Removed import of BadRequest: value became unused

Yutsuten committed:
Added 'await' to remaining 'Group.validateInvitations' functions

paglias commented on test/api/v3/unit/models/group.test.js:
```diff
@@ -556,47 +546,47 @@ describe('Group Model', () => {
 
-        expect(function () {
-          Group.validateInvitations(uuids, emails, res);
+        expect(async function () {
```
these `expect()...to.not.throw()` aren't working correctly with `async` functions. If you change the code to

```
      it.only('does not throw an error if emails are passed in and uuids are an empty array', async () => {
        expect(async function () {
          throw new Error('error');
          await Group.validateInvitations([], ['user1@example.com', 'user2@example.com'], res);
        }).to.not.throw();

        expect(res.t).to.not.be.called;
      });
```

the test passes even if it should not.

But you don't need to expect the function not to throw because if an error is thrown inside an async function the function and the test fails. So this will work

```
      it.only('does not throw an error if emails are passed in and uuids are an empty array', async () => {
        await Group.validateInvitations([], ['user1@example.com', 'user2@example.com'], res);
        expect(res.t).to.not.be.called;
      });
```

Yutsuten replied:
Makes sense. I'll fix it.

paglias commented on test/api/v3/unit/models/group.test.js:
```diff
@@ -460,73 +457,67 @@ describe('Group Model', () => {
+      it('throws an error if no uuids or emails are passed in', async () => {
+        await expect(Group.validateInvitations(null, null, res)).to.eventually.be.rejected.and.eql({
+          httpCode: 400,
+          message: 'Bad request.',
```
these messages shouldn't be the proper i18n strings like `canOnlyInviteEmailUuid ` in this case?

Yutsuten replied:
The message returned in the call to ``Group.validateInvitations`` is ``Bad request.``.

The only expectation about the call of this function is that it will fail with Bad Request. The reason of fail (``canOnlyInviteEmailUuid``) is on the ``res`` variable.

So, I can't see a way to do what you're requesting. Sorry.

paglias replied:
Yes, I'm wondering why the message is `Bad request` and not the real reason, it should not be `Bad message`. Does the test fails if `message: 'Bad request.'` becomes `message: shared.i18n.t('canOnlyInviteEmailUuid')`? 

I can't see where `Bad request.` as a message would come from. It should be like this https://github.com/HabitRPG/habitica/pull/8589/files#diff-ee4d06fc0df454c1e75431b5abb56e39R455

Yutsuten replied:
Yes, the test fails...
```
AssertionError: expected { Object (name, httpCode, ...) } to deeply equal { Object (httpCode, message, ...) }
+ expected - actual

 {
   "httpCode": 400
-  "message": "Bad request."
+  "message": "Can only invite using uuids or emails."
   "name": "BadRequest"
 }
```

paglias replied:
Ok, the `Bad request.` string seems to come from https://github.com/HabitRPG/habitica/blob/develop/website/common/script/libs/errors.js#L31 and it's used as a default when the message isn't provided and it happens because the `res` used in tests (https://github.com/HabitRPG/habitica/pull/8589/files/414e316e832f398b9a695aeb20f9223aa2f556f8#diff-548695dbee775dda2626c88e15cbcec6L459) doesn't return anything but it's just a stub. So false alarm :)

Yutsuten committed:
Fixed tests that would always success

Yutsuten commented:
If there is anything else that I should fix, please tell me! :)

paglias commented:
Merged :) I've also awarded your first contributor tier on Habitica. Really good work with the PR, thanks! 

Note that the PR being merged doesn't mean it's live on Habitica, it'll go there in a few days when it's also merged into the `release` branch

Alys commented:
Just as a note for our reference, paglias has pointed out that party invitations made by email to users outside of Habitica are not stored anywhere within Habitica, so they’re not considered when calculating the limit. This would allow parties to grow beyond 30 but it's unlikely to be a large problem and not worth the extra code and database storage that would be needed to keep track of emailed invitations.

paglias commented:
I've made two small changes in the group view https://github.com/HabitRPG/habitica/commit/daf9421f4f9e3b1ad15a675c9e06c69e2c57f7e8

They make sure the invite button isn't hidden in guilds and completely hide the panel used for the members limit message in guilds

Yutsuten commented:
Oh, those are very big mistakes of mine... I need to be more careful in my tests...
